server/README.md

```diff
@@ -0,0 +1,236 @@
+In this case, if Alice got a certificate with a timestamp more recent that Bob's certificate,
+then accepting Bob's certificate would break causality !
+
+### Example 2
+
+- At T1 the server receive block create from Alice in realm R1
+- At T2 the server receive a certificate unsharing with Alice realm R1 with timestamp T0
+
+In this case, and considering the server has accepted the unsharing certificate,
+any client accessing the block before T2 will consider everything is fine, while
+any client accessing the block after T2 will reject it given it appears to have
+been created after Alice lost her access to the realm !
+
+## Causality & concurrency model
+
+From the server point of view, there is basically four different kinds of data:
+
+- Certificates
+- Blocks
+- Vlobs
+- Invitations
+
+### Certificates
+
+Any change done by a client that should be visible by the server is provided as a
+certificate (i.e. a certificate is an unencrypted document signed by the client's device).
+
+A very simple way to ensure causality and concurrency on certificates would be:
+
+- Have a single big lock in the server that must be taken before any attempt to
+  add a certificate (handles concurrency).
+- Ensure the new certificate's timestamp is newer than any existing certificate
+  (handles causality).
+
+However this is pretty wasteful given most certificates are unrelated (e.g. certif A
+add a new user, certif B does a key rotation in a realm). So instead we reduce the
+size of the lock by introducing the concept of topic.
+
+Each type of certificate is part of one of the following topics:
+
+- `common`
+- `sequester`, which depends on `common`
+- `realm`, which depends on `common` and uses the realm ID as discriminant
+- `shamir_recovery`, which depends on `common` and use the ID of the user to recover as discriminant
+
+So the actual rule used in the server is:
+
+- Have lock in the server for each couple topic + discriminant
+- Take the lock for the topic + discriminant couple the certificate we want to add
+  belongs to AND to other topics our topic may depend on (handles concurrency).
+- Ensure the new certificate's timestamp is newer than any existing certificate
+  in the topics we have locked (handles causality).
+
+> Note: Also organizations are totally isolated from each other on the server, so
+> in a way the organization ID is an aditional discriminant on each certificate topic.
+
+### Blocks
+
+Blocks are encrypted piece of data that belong to a given realm and cannot
+be modified.
+
+This makes concurrency & causality pretty simple:
+
+- Each block is considered independent, hence there is no need for lock dedicated to blocks.
+- A block is identified by a client-provided UUID, then there is no risk for double-creation
+  under concurrency (a trivial unicity check on insertion prevents this).
+- Block creation date must be more recent than the last certificate in `common` topic and
+  in the `realm` topic corresponding to the block's realm.
```

Comment:
This causality check is only treated one way (checking the topic when creating a block). The opposite is not mentioned in the above section.

Reply:
Also this section does not mention locking: is the `common` topic lock taken why the block is uploaded?